### PR TITLE
Small improvements of completion functions

### PR DESCRIPTION
**Pull request by geraldus**
As a user of company mode I see errors in mini-buffer very often:

> Company: An error occurred in auto-begin
> Wrong type argument: stringp, nil

Small investigation lead me to proposed changes

### CODE REVIEW

**Comment by gracjan** [comment]
Is there a way to trigger this `nil` is some kind of test?

**Comment by geraldus** [comment]
@gracjan didn't understood your question well.  I've faced the situation, when GHCi fails to start:

> The following modules are present in multiple packages:
 * Internal.API (in solitaire-ghcjs-simple, ghcjs-perch-canvas)
 * Internal.FFI (in solitaire-ghcjs-simple, ghcjs-perch-canvas)
 * Internal.Type (in solitaire-ghcjs-simple, ghcjs-perch-canvas)

In this case REPL process is alive, and not marked as restarting and all requests to REPL returns `nil`.  As a consequence, response parser passes this `nil` value to `split-string` function which leads to error.  I have no strong opinion what to do with `nil`-ish responses.

**Comment by geraldus** [comment]
Oops.  I've found another one bug, force-pushed new changes.

TLDR: first fixes for static completion function added condition case for general type of prefix with nil value associated as result.  This was wrong, because instead of just `nil` value it returned a list of prefix boundaries positions and `nil`, making impossible other backends to proceed.

**Comment by purcell** [comment]
+1 for these fixes: `completion-at-point` in company is very broken without them. I was about to submit something like 4c6b3ffa6b295d3d9ae7cbacfef6a68812d021c7 myself.

**Comment by geraldus** [comment]
@purcell by the way, do you have some special setup for company mode and haskell?

**Comment by purcell** [comment]
Not really -- I mostly rely on `company-ghci` and `company-capf` in `haskell-mode`.

**Comment by geraldus** [comment]
@purcell I have found useful to override `company-backends` in haskell buffers, e.g.
```el
(set (make-local-variable 'company-backends)
       (append '((company-capf company-dabbrev-code) company-dabbrev)
               company-backends))
```
This combines completions from haskell mode with dabbrev-code completions outside strings and comments (useful for local identifiers defined in where and let blocks) and provides dabbrev completions otherwise.  I want to put it in docs.

**Comment by purcell** [comment]
@geraldus Good tip, thanks.
